### DISCUSSION TITLE
Backpack and Multi tenancy

### DISCUSSION THREAD

**Discussion by exorgroup** [Laravel-Backpack/community-forum#733] [Show and Tell]
I am implementing backpack in a multi tenancy application specifically using https://tenancyforlaravel.com/.  I know this is uncharted territory for me and may be for many but would like to share my experience here and may be receive some feedback to improve it 

So here is my two cents to this:

**Step 1:**
To install make sure you have a users table in the main database where the domains and tenants tables are - it is required for initial intall.

**Step 2:**
Once it installs it will log against the main table - change file `config\backpack\base.php`  by 
- commenting the line `'web_middleware' => 'web'`
- uncomment the section underneath it and add 
```php
         \Stancl\Tenancy\Middleware\InitializeTenancyByDomain::class,
         \Stancl\Tenancy\Middleware\PreventAccessFromCentralDomains::class, 
```

**Step 3:**
Change file `app\Http\Kernel.php` by adding the following references to `middlewareGroups`:
```php
         \Stancl\Tenancy\Middleware\InitializeTenancyByDomain::class,
         \Stancl\Tenancy\Middleware\PreventAccessFromCentralDomains::class, 
```
This allows you to log in against each tenant database and not the main database.

**Devtools**

I have started to use this and have no issues in generating most of the steps.  The only step that needs to be altered is the migration stage. **Do not migrate with devtools as it will migrate to the main database.**  This I step in manually shifting the migrations to the tenancy folder as described in the Stancl Tenancy documentation and perform a `tenancy migrate` instead or `artisan migrate`.  The rest is per devtools documentation and works fine.  I suppose that the migrate can be sorted as well by adjusting the code somewhere but did not look into that.

So far it works and have everything being generated in a multi tenant environment - if I broke something with these actions please let me know and potentially how it can be fixed but so far it seems to work good.

C Ya and regards to all

**Comment by tabacitu** [comment]
This. Is. AWESOME!

Thanks a lot for sharing your findings about this @exorgroup ! I'm am 100% this will help someone someday. I do the same - when I do something that's out-of-the-ordinary I document it. For myself and possibly others.

I've taken the liberty of editing your post slightly, added `highlighting` and **bold** to make it easier to read. But other than that, it reads really really well and I am glad there's a resource for people trying to do the same thing.

Moved to "Show and tell" since I think it's more appropriate there.

Thanks again for sharing this!

**Reply by exorgroup** [reply]
@tabacitu Thanks I appreciate - this is the nicety of a community - we learn from each other

**Comment by simke92** [comment]
Hi,

Over the past few days, I've been working on implementing multi-tenancy in my Backpack application, which involves using multiple databases and identifying tenants based on either subdomains or paths. When working with a fresh Laravel application, I followed the tenancy documentation, and everything worked smoothly with Breeze. I was able to successfully log in/register for both central and tenant databases and retrieve the correct user information.

However, I've encountered some challenges when trying to integrate Backpack into the setup. I suspect that I might not have a clear understanding of how to use middlewares and routes correctly, as I'm facing various issues. For instance, attempting to log in as a tenant with a central user, or trying to access any CRUD route, results in a "419 PAGE EXPIRED" error.

I would greatly appreciate it if anyone could provide a concise guide on how to effectively use these two excellent packages together. Specifically, I would like to know which route file should utilize the tenant middleware and which should use the central middleware.

Thank you in advance for your assistance!

**Reply by exorgroup** [reply]
You cannot log into a tenant with a central user.  To log in the tenant, it has to be a tenant user.  Central users log into central database. This is true irrespective of backpack or not.  It is how multitenancy is designed and it makes sense as you do not want your central user peaking into tenant's data but just manage the system when needed.  If you allow central user to peak into tenant's data than you may have some legislative issues.

I have managed to implement multi tenancy with backpack with good level of success and stability.  All the information has been placed in this article with a step by step guide what to alter and how it works.  Please note it is still works on progress as I am building the tenant part first and than will create all the tools for the central management later for my application, so I know with a good level of certainty that the tenant part routes well but cannot state the same for the central management.

Now for the 419 Page Expired error that look like it is logged out through some time out session.  I get that when it times out based on my settings than I have to log in again.  There I notice that the re-routing not always work right but when I click on an item on the menu with the proper routing it works like a charm.

**Comment by JackVanson** [comment]
Hi @exorgroup Thank you so much for this. I am looking to do the exact same. I have Tenancy For Laravel and was trying to use Nova with very little success when trying to create a form builder. Coming across Backpack seems to be the best option! 

I would love to hear how you have got on with your development. Anything I should watch out for, any tips and tricks. I would really appreciate any advice that you can share. 

Many thanks

Jack

**Reply by exorgroup** [reply]
If you are referring to multi-tenancy, the only advice I can give is to test well.  Not easy to get it right but I hope that my explanation is easy to follow.  If the first time trying this, I really recommend try on a blank project.

With reference to Backpack - it is an awesome platform.

**Comment by siberfx** [comment]
@tabacitu I used to build SaaS applciations with backpack but there were some issues that needs to be adjusted by backpack himself, especially for the BL when we reach out the model instances.
instead of using Model::class, we should be able to inject the model $connection parameter neither in controllers nor it should directly use the Model himself like new Model;  also to define the database credentials there must be a a way of implementing all URI's parameters like domain=blabla  to define current database :) I hope you understand what I meant

**Comment by siberfx** [comment]
used backpack for admin panel and I was able to pass tenant related data in it, but its not so stable as expected. especially when updating creating data the redirects are giving pain in the a$$ :)

**Reply by JackVanson** [reply]
How did you manage all the routes. I seem to be having a lot of issues with the many route files.

**Reply by siberfx** [reply]
well if you set your logged in user with the tenants table within a relationship, then you can use for the logged in users domain key and set new database credentials in the checkIfAdmin Middleware. its one of the way.

**Comment by martincarvalho** [comment]
@exorgroup since Laravel 11 doesn't have `Kernel.php` anymore, how should I proceed with step 3? I'm trying to make this work on a brand new installation of Laravel 11 + Backpack with no luck

**Reply by emjayess** [reply]
It would seem reasonable to adopt or attempt one of these middleware registration approaches, from the docs:
* https://laravel.com/docs/11.x/middleware#registering-middleware

Alternatively, I suspect you could also just try adding `Kernel.php` back where it used to be, wire it up the way one would in Laravel 10, and assert or expect that might also still work that way. I count myself among folks who think the Laravel team went just a tad too aggressive just to "lean down" the framework, and did not take into consideration how much conflicting or forked documentation online will now just end up on a path of confusion. :/

**Reply by siberfx** [reply]
if you guys are still interested I have figured out and I can share the way I did with anyone who's interested in

**Reply by bahaa-almahamid** [reply]
> if you guys are still interested I have figured out and I can share the way I did with anyone who's interested in

Hi, do you have a link to a repo with the solution? 
thanks
